### PR TITLE
Strings: add ByteVector

### PR DESCRIPTION
**Pull request by liyishuai**
**Kind:**  feature

- [x] Entry added in CHANGES.

### CODE REVIEW

File: CHANGES [hunk 1]
```diff
@@ -140,6 +140,11 @@ Standard Library
 - There are now conversions between [string] and [positive], [Z],
   [nat], and [N] in binary, octal, and hex.
 
+- Added `ByteVector` type that can convert to and from [string].
+
+- Added `uncons` and `splitAt` functions in over vectors, and proved some lemmas
```
Comment (Zimmi48) [inline CHANGES]: "in over"?

**Review by Zimmi48 — APPROVED** [review]
Approving the trivial doc update part.

**Comment by liyishuai** [comment]
I have no idea why my latest changes caused that [error](https://gitlab.com/coq/coq/-/jobs/93701292). Proposing them in another PR.

**Review by herbelin — COMMENTED** [review]
Looks good to me, but can you comment a bit more in passing on your motivation / use cases?

**Comment by liyishuai** [comment]
`ByteVector` provides convenience in processing structured bytes. For example, to parse a binary file of a certain format, we can convert the string into `ByteVector`s. For numeric fields, we can convert it to `Bvector` and get its value in `N`.

**Comment by herbelin** [comment]
Actually, I'm curious about whether QuickChick would be a client of these translators, or if the motivation comes from another project?

**Comment by liyishuai** [comment]
The motivation comes from [coq-http2](https://github.com/liyishuai/coq-http2) that specifies the behavior of  HTTP/2 servers. The encoder and decoder use `ByteVector` to represent the frames, whose fields are mostly of fixed size.

**Comment by ejgallego** [comment]
I understand @herbelin is the right developer to take this PR.

**Comment by ejgallego** [comment]
Thanks @liyishuai !

**Comment by liyishuai** [comment]
I'm confused about the failure in [ci-plugin-tutorial](https://gitlab.com/coq/coq/-/jobs/109436852). Did I break something?

**Comment by Zimmi48** [comment]
Nope, it is failing everywhere unfortunately :disappointed:

**Comment by ejgallego** [comment]
> Nope, it is failing everywhere unfortunately

If we don't have an ETA on @ybertot return, let's do a PR moving the ref to a sane place.

**Comment by maximedenes** [comment]
> If we don't have an ETA on @ybertot return, let's do a PR moving the ref to a sane place.

He'll be back on Monday, but then there's the WG, and he's on holidays for a few days. So I'd say something like Oct. 29.

**Comment by herbelin** [comment]
@liyishuai: would it work to use a `fun 'pat` construction rather than the sequence of `uncons`?

Otherwise, assigning.

File: theories/Strings/ByteVector.v [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  | [] => []
+  | Ascii b0 b1 b2 b3 b4 b5 b6 b7::v' =>
+    b0::b1::b2::b3::b4::b5::b6::b7::to_Bvector v'
+  end.
+
+Fixpoint of_Bvector {n : nat} : Bvector (n * 8) -> ByteVector n :=
+  match n with
+  | 0 => fun _ => []
+  | S n' =>
+    fun v =>
```
[the remark this answers is not in the document]
Comment (herbelin) [reply theories/Strings/ByteVector.v]: Yes, indeed, type-checking is not powerful enough to accept it. Too bad. Let's keep it as it is.

**Comment by herbelin** [comment]
Shall merge on Monday late if no other comments.

**Pull request merged**